index.bs

```diff
@@ -2073,6 +2151,69 @@ Methods</h4>
 		<div>
 			<em>Return type:</em> {{Promise}}&lt;{{undefined}}&gt;
 		</div>
+
+		: <dfn>setSinkId((DOMString or AudioContextSinkCategory) sinkId)</dfn>
+		::
+			Sets the ID of the output device if it is permitted to play audio. When
+			this method is invoked, the user agent MUST run the following steps:
+
+			<div algorithm="AudioContext.setSinkId()">
+				1. Let |document| be the current settings object's
+					[=associated Document=].
+
+				1. If |document| is not allowed to use the feature identified by
+					<code>"speaker-selection"</code>, return [=a promise rejected with=] a
+					new {{DOMException}} whose name is "{{NotAllowedError}}".
+
+				1. Let |context| be the {{AudioContext}} object on which this method was
+					invoked.
+
+				1. Let |sinkId| be the method's first argument.
+
+				1. If |sinkId| is equal to |context|'s {{AudioContext/[[sink ID]]}}
+					internal slot, return a promise resolved with {{undefined}}.
+
+				1. Let |p| be a new promise.
+
+				1. Run the following substeps:
+
+					1. If |sinkId| is the empty string, reject |p| with a new
+						{{DOMException}} whose name is "{{NotFoundError}}" and abort these
+						substeps.
+
+					1. If |sinkId| does not match any audio output device identified by
+						the result that would be provided by
+						{{MediaDevices/enumerateDevices()}}, reject |p| with a new
+						{{DOMException}} whose name is "{{NotFoundError}}" and abort these
+						substeps.
+
+					1. If the device specified by |sinkId| is not the user agent's
+						default device or is not permitted to play audio through, reject |p|
+						with a new {{DOMException}} whose name is {{NotAllowedError}} and
+						abort these substeps.
+
+					1. Stop playing audio out of the current output device.
+
+					1. Run the following substeps to prepare an audio output device for
+						rendering:
+
+						1. Set context's {{AudioContext/[[sink ID]]}} internal slot to
+							|sinkId|.
+
+						1. If {{AudioContext/[[sink ID]]}} is <code>"{{AudioContextSinkCategory/none}}"</code>, acquire system
+							resources to render the audio graph without an actual audio
+							output device. Abort these substeps.
+
+						1. If {{AudioContext/[[sink ID]]}} is not the empty string, acquire system
+							resources to render the audio graph with the underlying audio
+							output device identified by {{AudioContext/[[sink ID]]}}.
```

Comment:
What would be cool here is to suspend the graph, so that authors can use `onstatechange` to determine when the audio is flowing again, because it can take some time for this to happen, and the only option is to resort to polling `currentTime`.

We should be able to reuse the `suspend` and `resume` methods. This has also the advantage of making it clear that this will happen in between two render quantums.

Reply:
Not sure I am following here - so setSinkId() will cause to suspend, and then automatically resume later? So the user code use the `onstatechange` signal for the renderer's readiness?

I think `onsinkchange` might be a better signal for it.

I guess this should be another topic for discussion?

Reply:
Does `onsinkchange` fire if the default audio output changes? i.e. you're just playing on the default audio output device, and it's being unplugged, so that the device has changed? It would be useful: something like the sample-rate or optimal render quantum size could be different, authors might want to rebuild their audio graph.

Other than that, my proposal was that:
- The promise from `setSinkID(...)` is resolved after the audio is flowing again (i.e. it's resolved after the operation has succeeded)
- The `AudioContext.state` is `"suspended"` while the graph is not being rendered because the underlying audio output device is changing
- `onstatechange` fires normally.

It would go like this:
- `setSinkID(...)` is called with a valid device identifier, returning a promise `p`
- The rendering is paused, possibly after the current rendering quantum
- `"statechanged"` is fired, the state is now `"suspended"`
- The old device's stream is closed
- A new system-level audio stream is opened, on the new device
- Some time passes, opening the device can take time
- The new stream initialization has finished
- Queue a task to fire `"sinkchange"` the `sinkID` is now the new `sinkID`, resolve `p`
- Queue a task to fire `"statechange"`, the state is now `"running"`

Reply:
I'll consider this change in the next round of editing.